### PR TITLE
Added vendor dir for js/css libs; Documented sources (#1484)

### LINKED ISSUE #1484: Unsourced/undocumented libraries; missing license files; and other issues
I noticed that there are some compiled javascript files that don't list an upstream source. I was able to find some in public/assets/librejs/librejs.html, but seem to be missing the following:

dropzone-4.2.0
jquery.datetimepicker-2.4.5
jquery.are-you-sure
pdfjs-1.4.20
codemirror-5.17.0
autolink.js
gitgraph
draw
font-awesome-4.6.3
octicons-4.3.0
public/less/?
public/../github-min.js?
public/themes/*fonts*?

Could we get librejs.html updated with this information, please? Also, if there have been any modifications of these files, have they been documented anywhere?

While I was looking through public/img/, I saw images such as slack.png, openid-x.png, and github.png that I would assume are proprietary and don't provide sources. The emoji/ directory also seems to have a large number of images that probably have sources that I missed.

I would like to locate a license grant for any proprietary images or see documentation pointing at the upstream source and license. Ideally, I'd like some script that could be used without requiring network access that can reproduce the images in public/img/.

In a perfect would, the packaging process would strip public/ and rebuild everything in it, but the above is a bare minimum if I have any hope of getting this into Debian.

Thanks!!

### PR DESCRIPTION
**Pull request by MTecknology**
This PR resolves #1484 and moves vendor'ed javascript (and other) libraries to public/vendor/. It also moves the documentation to vendor/librejs.html and includes documentation for both js and non-js libs.

With this commit, it is easy to prove only free/open-source libraries are being used by gitea, which helps meet GNU ethical repository criteria (#1524).

### CODE REVIEW

**Comment by MTecknology** [comment]
Note: before merging, librejs.html still needs to be cleaned up.

**Comment by sapk** [comment]
LGTM
Just, why do you choose to put all under /plugins ? Why not keep some under sub-folder /libs like jquery, vue or semantic ?

**Comment by MTecknology** [comment]
@sapk 

There were times where the CSS components weren't kept with the JS bits, but that wasn't consistent. I moved each lib into it's own directory partly to add consistency; I also wanted to be able to include a license file for each vendor'ed library to clearly identify each project as a separate work and clearly show the license terms of that source (which also helps meet license terms).

I kept them under plugins/ instead of breaking them into the other sub-directories because I wasn't able to discern the logic to the current layout.

Example:
```
jquery-1.11.3 was under public/js/
jquery.are-you-sure was under public/js/libs/
two other jquery libs were under public/plugins/<lib-name>/

public/js/libs/
    clipboard-1.5.9
    emojify-1.1.0
public/plugins/
    highlight-9.11.0
    simplemde-1.10.1
```

Even in the source, I couldn't find a reason why they should be considered different. Sorry, I didn't mean to get long-winded or ranty... just wanted to thoroughly explain my logic. :)

**Comment by sapk** [comment]
@MTecknology ok not a problem just to know if there were any reason. But effectively there wasn't before also ^^. Maybe we should think of using versionning for web ressources like npm or bower in the future. (not sure everyone will aggree with me ^^). still LGTM

**Comment by lafriks** [comment]
@sapk in the future we should move to webpack to have single js file

**Comment by MTecknology** [comment]
@lafriks I agree! It'd be nice to have a nice json file (or two) with a pretty script to update all libs and run unit testing. If we had that, librejs.html could be turned into a dynamic file that's built by reading existing json files. (sounds great for a lot of reasons)

For now, though, it'd be super if this could make it into 1.2.0. It'd make my life a whole super duper lot easier. :)

(P.S. I actually don't care that much about the last commit. I hope to never touch that librejs plugin again... sticking w/ noscript. I just happened to be "in the neighborhood.")

File: public/assets/librejs/librejs.html [hunk 1]
```diff
@@ -1,84 +0,0 @@
-<!DOCTYPE html>
```
Comment (bkcsoft) [inline public/assets/librejs/librejs.html:1]: Why has this file changed?

**Review by bkcsoft — CHANGES_REQUESTED** [review]
A part from the question. There's no file listing which versions we have. This _have_ to exist for this to be merged. I don't care if it's a real `package.json/yarn.lock` or just a plain-text file listing them, but it absolutely have to exist.

**Comment by bkcsoft** [comment]
> don't care if it's a real `package.json`/`yarn.lock` or just a plain-text file listing them

With this I mean that for _this_ PR I don't care, any updates in the future should use yarn though ;)

File: public/assets/librejs/librejs.html (hunk 1 above)
Comment (sapk) [reply public/assets/librejs/librejs.html:1]: It has moved under public/vendor/librejs.html

**Comment by sapk** [comment]
@bkcsoft we could do that in an other PR.

**Comment by sapk** [comment]
And in fact all the version are listed in archive link of librejs.html

**Comment by bkcsoft** [comment]
@sapk Why I hate PRs that touch 1.3k files 😂

**Comment by sapk** [comment]
Me too but in this case it is mostly rename.

**Comment by bkcsoft** [comment]
@sapk right, my point was that we need a file `public/vendor/VERSIONS` that just lists all versions. Since I had no idea that `librejs.html` had that, and therefore no one else knows that either 😉

**Comment by MTecknology** [comment]
@bkcsoft I think I [already responded](https://github.com/go-gitea/gitea/pull/2241#issuecomment-320394389) to most of those concerns. I agree, librejs.html should be dynamically generated from the output of a tool that can keep JS libs up to date. (hopefully choosing a tool that can output a "versions" file)

I also think that's something much better suited for a different PR. I tried to keep this PR to just file moves and documentation--it's big, but I tried for minimal. librejs.html wasn't listed as a move because the file contents changed almost entirely.

Also, in the future, it'd be nice to apply a template to the rendered page so it's not just a boring table but a themed page with ```<table>[...]</table>``` as the page body.

**Comment by strk** [comment]
LGTM

**Comment by strk** [comment]
This is ready to merge,  why waiting for 1.3.0 ? I think it is worth having this ib 1.2.0

**Comment by lunny** [comment]
It's not tested well. I don't think move it to v1.2 is a good idea.

**Comment by strk** [comment]
@lunny 1.2.0 could be the first version with a proper Debian package, but I've understood a Debian package depends on this PR, for licensing issues.

#2285 was more dangerous to put in 1.2.0, if you ask me, as I suspect it's what broke the UI (#2331)

**Comment by MTecknology** [comment]
@lunny This is a large PR, but I would encourage you to review what is being changed.

1. Moved 3rd party libs to designated directory
2. Updated path references to match
3. Created/updated library documentation
3a. Added missing LICENSE files
4. Fixed issues for librejs users

This PR addresses some licensing concerns that, as @strk mentioned, are blockers for other bugs--one of those being a proper Debian package. I intentionally kept this changeset as small as possible to prevent this problem. I don't know what additional testing can be performed; if you tell me, I'll do my best to make it happen.

@bkcsoft You still have "changes requested," but I do not see what changes you are requesting. Could you please tell me so that I can make them?

**Comment by bkcsoft** [comment]
@MTecknology A file listing packages and versions. (no, `librejs.html` does not count 🙂 )

**Comment by MTecknology** [comment]
@bkcsoft Your comment was about the file being removed. The file was not
removed, it was relocated and modified.

I already commented on the inclusion of an alternate option. It's a cool
idea. This PR is already big enough, isn't it?

It's a really big PR and, as I mentioned, I tried hard to keep it minimal.
I tried to keep out functional changes and keep it to documentation and
cleanup.

I disagree very much that this should be held up to wait for an entirely
new and undiscussed feature.

**Comment by bkcsoft** [comment]
@MTecknology It's not a new feature, the version were tracked by the folder-names (`jquery-1.1.3` etc), but now that information is removed and needs to be added back in some form. And as I've said, it does _not_ have to be yarn/npm, just a straight up flat file listing deps and versions...

**Comment by strk** [comment]
Is a machine readable format you are after, @bkcsoft ?

**Comment by lunny** [comment]
I found some links broken in librejs.html master version. If this PR will also fix that?

**Comment by MTecknology** [comment]
@lunny Yes...

The [first commit](https://github.com/go-gitea/gitea/pull/2241/commits/9f357b5622393792980ad18237e7d80a7707484f) in this PR:
> This commit [moves] vendor'ed plugins into a vendor/ directory and document[s] their upstream source and license in vendor/librejs.html.

I'll say this again... very little actually changes in this PR. It's documentation and licensing crap. Please, look at what [files changed](https://github.com/go-gitea/gitea/pull/2241/files) and what file changes took place. I tried very hard to make this PR very easy to review for these exact reasons... I thought I'd succeeded... :(

File: public/assets/librejs/librejs.html (hunk 1 above)
Comment (MTecknology) [reply public/assets/librejs/librejs.html:1]: [As requested](https://github.com/go-gitea/gitea/pull/2241#issuecomment-320909454) => the [VERSIONS](https://github.com/go-gitea/gitea/pull/2241/commits/926cb40203569545940cde0deb20aefc6897a34e) file

**Comment by bkcsoft** [comment]
@MTecknology My main quarel with having them listed in `librejs.html` is because that file is not where I'd expect to find that information. Since you've now added a `VERSION` file this LGTM 🙂

**Review by bkcsoft — APPROVED** [review]

**Comment by lofidevops** [comment]
Awesome!

**Comment by lunny** [comment]
And If you want to this in v1.2, please send a back port PR to branch release/v1.2

**Comment by gsantner** [comment]
I'm on current HEAD and there are things broken. I think it's related to this one

unbenannt

**Comment by daviian** [comment]
Furthermore integration tests are failing because of moved librejs file.

**Comment by strk** [comment]
Is drone not running integration tests ?

**Comment by daviian** [comment]
@strk @tboerger mentioned that integration tests currently run only on master branch...

**Pull request merged**